Pass manager refactoring: add passmanager module

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This PR adds new `qiskit.passmanager` module and reorganizes `qiskit.transpiler`. New module implements a generic pass manager which is not sensitive to object types (input, output, IR) under transformation, and only provides a management for pass execution.

1 of 2 PRs. Pass manager internals will be refactored in the follow up.

### Details and comments

The internal logic is almost as-is. One visible change is in `qiskit.passmanager.passrunner`. This provides `BasePassRunner` which is a superclass of `RunningPassManager`. New base class implements
- `BasePassRunner._to_passmanager_ir`
- `BasePassRunner._to_target`
- `BasePassRunner._run_base_pass`

which are all abstract methods. These are type sensitive and intended to be implemented by a subclass for particular data types, e.g. for existing pass manager, input: QuantumCircuit, passmanager_ir: DAGCircuit, output: QuantumCircuit. We plan to introduce another compiler for the pulse module.

The new methods

- `BasePassRunner._run_pass_generic`

implements data type agnostic flow control. This is a common machinery for every subclass.

nkanazawa1989 commented on qiskit/passmanager/propertyset.py:
```diff
@@ -10,8 +10,11 @@
+
+
+This is sort of shared memory space among passes.
+"""
```
Makes sense. Done in [729b174](https://github.com/Qiskit/qiskit-terra/pull/10124/commits/729b1749a26cb970321235bd64ac3ed11dd5ff7b).

nkanazawa1989 commented on qiskit/passmanager/__init__.py:
```diff
@@ -0,0 +1,107 @@
+Note that the input type and output type don't need to match.
+
+Execution of transform pass is managed by the :class:`.FlowController`,
+which is initialized with a set of transform passes and provides an iterator of passes.
```
I don't plan to implement transform and analysis pass in the base passmanager module, e.g. for very simple use case probably this might be overkill. In this sense I avoided to distinguish them in this documentation. I remove _transform_ and just call them passes.

nkanazawa1989 commented on qiskit/passmanager/passmanager.py:
```diff
@@ -0,0 +1,264 @@
+logger = logging.getLogger(__name__)
+
+
+class BasePassManager(ABC):
```
Actually a developer needs to attach `BasePassManager.PASS_RUNNER` to make it work. In this base class `BasePassRunner` is just added for convenience of type hint, but the base class doesn't provide IR conversion logic. One could also do abstract class property, but I don't like class property. 

Alternatively, we can keep `BasePassManager._create_running_passmanager` method and make it abstract. However, writing and overriding this method just for difference of pass runner type seems awkward (we can remove this method in the followup PR #10127 ). 

The another place where the pass runner type is necessary is [run](https://github.com/Qiskit/qiskit-terra/pull/10124/files#diff-9432898faf9979ca2ac7448bfd24071382f544b42aac863790fe965d1dd1f3d2R216-R218) method. We need to dispatch the `_run_single_circuit` when input is a single entry, but we cannot safely count entry number because input program is usually iterable, e.g. `QuantumCircuit` and `ScheduleBlock`.

nkanazawa1989 commented on qiskit/transpiler/passmanager.py:
```diff
@@ -507,7 +382,15 @@ def remove(self, index: int) -> None:
+        if isinstance(_pass_sets, dict):
+            _pass_sets = [_pass_sets]
+        new_passmanager._pass_sets = _pass_sets
+        return new_passmanager
```
No. In the current implementation the renewed class instance type is hard-coded like this. 

https://github.com/Qiskit/qiskit-terra/blob/cff51cea9bd136534154fca4a5f29631022500ed/qiskit/transpiler/passmanager.py#L149-L163

This will always return `qiskit.passmanager.PassManager`, regardless of subclass. This is not the case of new base class, and it must return `self.__class__` instead of hard-coded `PassManager`. This change breaks the add dunder method of the `StagedPassManager` because it just inherits from the new base class, which tries to return `StagedPassManager`.

nkanazawa1989 commented on releasenotes/notes/add-passmanager-module-3ae30cff52cb83f1.yaml:
```diff
@@ -0,0 +1,34 @@
+    :class:`~qiskit.passmanager.DoWhileController` are moved
+    from :mod:`qiskit.transpiler.runningpassmanager`
+    to own module :mod:`qiskit.passmanager.flow_controllers`.
+  - |
+    New error baseclass :class:`~qiskit.passmanager.PassManagerError` is introduced.
+    This will replace :class:`~qiskit.transpiler.TranspilerError` raised in the
+    pass handling machinery. The TranspilerError is now only used for the errors
+    related to the failure in handling the quantum circuit or DAG circuit object.
+    Note that the TranspilerError can be caught by the PassManagerError
+    because of their class hierarchy. For backward compatibility,
+    :class:`qiskit.transpiler.PassManager` catches PassManagerError and
+    re-raises the TranspilerError. This error replacement will be dropped in future.
```
Yes, but I think this will be in the next release according to our deprecation policy. As I don't change any test, there is no loss of backward compatibility.

mtreinish replied:
Ok I guess an upgrade note is fine for this. I was more just trying to figure out whether this was a call to action for users catching `TranspilerError` before or whether we should document it somewhere else. The only other thought I had was maybe as an `other` note but this seems like a better place.

nkanazawa1989 commented on qiskit/transpiler/passmanager.py:
```diff
@@ -541,3 +424,29 @@ def draw(self, filename=None, style=None, raw=False):
+    return wrapper
+
+
+for _name, _method in inspect.getmembers(PassManager, predicate=inspect.isfunction):
```
I didn't try detailed profiling and ASV benchmark, but likely it just causes a slight overhead at module loading (percentage is too small to be detected). Since this only wraps public methods (such as `.run`), performance regression during run time is not detected. 

![image](https://github.com/Qiskit/qiskit-terra/assets/39517270/abcfe061-a463-4609-af3f-78af1731818e)

If this causes a visible regression in the ASV, I'll open followup PR.

mtreinish commented on releasenotes/notes/add-passmanager-module-3ae30cff52cb83f1.yaml:
```diff
@@ -0,0 +1,34 @@
+    See upgrade notes for more details.
+upgrade:
+  - |
+    :class:`qiskit.transpiler.PassManager` is now a subclass of
+    :class:`qiskit.passmanager.BasePassManager`. There is no functional modification
+    due to this class hierarchy change.
+  - |
+    :mod:`qiskit.transpiler.propertyset` is moved to :mod:`qiskit.passmanager.propertyset`.
+  - |
+    Pass flow controllers :class:`~qiskit.passmanager.FlowController`,
+    :class:`~qiskit.passmanager.ConditionalController` and
+    :class:`~qiskit.passmanager.DoWhileController` are moved
+    from :mod:`qiskit.transpiler.runningpassmanager`
+    to own module :mod:`qiskit.passmanager.flow_controllers`.
```
I'm not sure we need upgrade notes for these anymore since there is nothing users need to do now. In 0.26 when we deprecate the legacy path we can document it then.

mtreinish reviewed (dismissed):
This LGTM now, thanks for the updates and sorry for the slow turnaround. I left a few comments about release notes but we can just adjust those as part of the release prep.

I was starting to review the code for the pass manager itself and was going to ask a lot of questions about specifics and things that looked really weird to me until I realized most of that was pre-existing and just migrated code. I'll save those reviews for the next PR in the series that refactors the internals

jlapeyre commented on qiskit/passmanager/__init__.py:
```diff
@@ -0,0 +1,107 @@
+* Converting the IR back to an output object.
+
+A single pass runner always takes a single input object and returns a single output object.
+Parallelism for multiple input objects is supported by the :class:`BasePassManager` by
+broadcasting the pass runner via multiprocessing is available by internally
+leveraging the :mod:`qiskit.tools.parallel_map` function.
```
This sentence is awkward to parse near the words "is available". Does the following capture what you intend?
```suggestion
Parallelism for multiple input objects is supported by the :class:`BasePassManager` by
broadcasting the pass runner via the :mod:`qiskit.tools.parallel_map` function.
```

jlapeyre commented on qiskit/passmanager/__init__.py:
```diff
@@ -0,0 +1,107 @@
+
+The base class :class:`BasePassRunner` doesn't define any associated type by itself,
+and a developer needs to implement a subclass for a particular object type to optimize.
+This `vail of ignorance` allows us to choose the most efficient data representation
```
At least three homophones for this word :roll_eyes: 
```suggestion
This `veil of ignorance` allows us to choose the most efficient data representation
```

jlapeyre reviewed (changes requested):
I left a couple of suggestions on documentation that should be reviewed.

jlapeyre commented on qiskit/passmanager/base_pass.py:
```diff
@@ -0,0 +1,80 @@
+
+    def __init__(self):
+        self.requires = []  # List of passes that requires
+        self.preserves = []  # List of passes that preserves
```
My best recollection is that this attribute has not been useful in practice. In fact, as far as I can see, it is not used at all in terra. In contrast, `requires` is used a bit. If it were not for backward compatibility, I imagine it would be removed. So I question whether it is a good idea to include this as a feature in the more abstract pass manager. Is it possible to define this attribute instead in `BasePass(CommonBasePass)` below?

mtreinish replied:
I think for this PR it's ok to preserve `preserves` as this is strictly for migrating the common code that already exists into a reusable module. There is a second PR #10127 built on top of this which aims to clean things up. I think we can discuss specifics like that there. I think it would be easier to view this PR as just a code migration for the ease of review. We're not locked into the api here until the final release anyway so there is time to change it post merge

nkanazawa1989 replied:
Thanks @jlapeyre , I agree with you. I also observed that options dictionary is carried around passes, 
https://github.com/Qiskit/qiskit-terra/blob/924702e9ddb3f6f0124bc8acb285ac3d1da12aff/qiskit/transpiler/runningpassmanager.py#L153
but eventually no one use this information (at least in Terra passes). However, I cannot decide what should be removed. I also want to cleanup and drop un-used features. Let's discuss carefully in #10127 . Once this is merged I'll rebase the PR for easier review.

jlapeyre replied:
Sure it's ok, as long as it doesn't fall through the cracks. Btw. I think one or two people wrote or told me that `preserves` was an idea copied from llvm without knowing whether it would be useful.

jlapeyre commented on qiskit/transpiler/basepasses.py:
```diff
@@ -13,61 +13,18 @@
-    Enforces the creation of some fields in the pass while allowing passes to
-    override ``__init__``.
-    """
+from qiskit.passmanager.base_pass import BasePass as CommonBasePass
```
This means we have both `passmanager.BasePass` and `transpiler.basepasses.BasePass`. This will likely cause a little confusion when searching for things in the source tree. But, probably a good IDE would help. Also, in cases like this, people argue that this is exactly why namespaces exist. I think leaving this as you have it might not be a bad choice, but i'ts worth considering changing it. If I had a good substitute name, I'd suggest it. I think using `passmanager.CommonBasePass` might make the `passmanager` module too verbose.

nkanazawa1989 replied:
That's fair point. Since we have another cleanup PR #10127 , we can also change the class name in that PR if we come up with better substitute name. I would prefer changing the `transpiler.basepasses.BasePass` to `CircuitBasePass` and adding an alias of `BasePass = CircuitBasePass`. Although this is a public class, the base class is immediately subclassed by transform and analysis pass. I assume impact to users is limited.

jlapeyre replied:
`CircuitBasePass` . yes this is a good idea.

jlapeyre commented on qiskit/passmanager/flow_controllers.py:
```diff
@@ -0,0 +1,153 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
```
This copyright notice should be updated.

jlapeyre commented on qiskit/passmanager/passrunner.py:
```diff
@@ -0,0 +1,249 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
```
Year should be updated.

jlapeyre commented on qiskit/passmanager/propertyset.py:
```diff
@@ -0,0 +1,24 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
```
Year should be updated

jlapeyre commented:
See also a previous PR #9163 that was closed in favor of this one.

nkanazawa1989 commented:
Thanks for the review. I just updated document in [ba6e698](https://github.com/Qiskit/qiskit-terra/pull/10124/commits/ba6e6985a8b632ef76b5f7f58ef19312402686c7) and rebased this PR. @jlapeyre Let's discuss BasePass naming and removal of unused feature in the followup PR #10127 .

nkanazawa1989 commented:
Thanks @jlapeyre for approval. After discussion I came up with another name _GenericPass_. The `passmanager.base_pass.GenericPass` (BasePass) doesn't implement any feature specific to a particular IR / transformation, and this make me think this is sort of generic class.

> Generic classes encapsulate operations that are not specific to a particular data type. The most common use for generic classes is with collections like linked lists, hash tables, stacks, queues, trees, and so on. Operations such as adding and removing items from the collection are performed in basically the same way regardless of the type of data being stored.

https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/generic-classes

mtreinish commented:
LGTM, thanks for all the updates. Lets continue the review of the specific internals of the passmanager in #10127
